src/harness/fourslash.ts

```diff
@@ -275,19 +272,35 @@ module FourSlash {
         private addMatchedInputFile(referenceFilePath: string) {
             var inputFile = this.inputFiles[referenceFilePath];
             if (inputFile && !Harness.isLibraryFile(referenceFilePath)) {
-                this.languageServiceShimHost.addScript(referenceFilePath, inputFile);
+                this.languageServiceAdaptorHost.addScript(referenceFilePath, inputFile);
             }
         }
 
-        constructor(public testData: FourSlashData) {
-            // Initialize the language service with all the scripts
-            this.cancellationToken = new TestCancellationToken();
-            this.languageServiceShimHost = new Harness.LanguageService.TypeScriptLS(this.cancellationToken);
+        private getLanguageServiceAdaptor(testType: FourSlashTestType):
+            { new (cancellationToken?: ts.CancellationToken, options?: ts.CompilerOptions): Harness.LanguageService.LanguageServiceAdaptor } {

```

Comment:
I'm not sure if this is actually better than `LanguageServiceAdaptorFactory` (which is kind of terrible on its own).